File: src/lib/api/comprobantes.ts
import { invoke } from "./client";

export type TipoComprobante = "ticket" | "a4";
export type TipoEventoComprobante = "impreso" | "pdf_generado";

export type Comprobante = {
  id: number;
  ventaId: number;
  numero: string;
  tipo: TipoComprobante;
  creadoEn: string;
};

export type ComprobanteEvento = {
  id: number;
  tipoEvento: TipoEventoComprobante;
  fecha: string;
};

/** Idempotente: si ya existe un comprobante de ese tipo para la venta, lo devuelve tal cual. */
export function obtenerOCrearComprobante(ventaId: number, tipo: TipoComprobante) {
  return invoke<Comprobante>("comprobantes_obtener_o_crear", { datos: { ventaId, tipo } });
}

/** Solo lectura: nunca genera un comprobante. */
export function listarComprobantesPorVenta(ventaId: number) {
  return invoke<Comprobante[]>("comprobantes_listar_por_venta", { ventaId });
}

export function registrarEventoComprobante(comprobanteId: number, tipoEvento: TipoEventoComprobante) {
  return invoke<void>("comprobantes_registrar_evento", { comprobanteId, tipoEvento });
}

export function listarEventosComprobante(comprobanteId: number) {
  return invoke<ComprobanteEvento[]>("comprobantes_listar_eventos", { comprobanteId });
}

File: src-tauri/src/commands/comprobantes.rs
use tauri::State;

use crate::db::AppState;
use crate::error::AppResult;
use crate::models::comprobante::{Comprobante, ComprobanteEvento, ObtenerOCrearComprobante};
use crate::services::comprobantes;

#[tauri::command]
pub async fn comprobantes_obtener_o_crear(
    state: State<'_, AppState>,
    datos: ObtenerOCrearComprobante,
) -> AppResult<Comprobante> {
    comprobantes::obtener_o_crear(&state.pool, datos.venta_id, &datos.tipo).await
}

#[tauri::command]
pub async fn comprobantes_listar_por_venta(
    state: State<'_, AppState>,
    venta_id: i64,
) -> AppResult<Vec<Comprobante>> {
    comprobantes::listar_por_venta(&state.pool, venta_id).await
}

#[tauri::command]
pub async fn comprobantes_registrar_evento(
    state: State<'_, AppState>,
    comprobante_id: i64,
    tipo_evento: String,
) -> AppResult<()> {
    comprobantes::registrar_evento(&state.pool, comprobante_id, &tipo_evento).await
}

#[tauri::command]
pub async fn comprobantes_listar_eventos(
    state: State<'_, AppState>,
    comprobante_id: i64,
) -> AppResult<Vec<ComprobanteEvento>> {
    comprobantes::listar_eventos(&state.pool, comprobante_id).await
}
